library/Translate/src/Juvix/Desugar/Passes.hs

```diff
@@ -0,0 +1,382 @@
+translateDo :: Sexp.T -> Sexp.T
+translateDo xs = Sexp.foldPred xs (== ":do") doToBind
+  where
+    doToBind atom sexp =
+      Sexp.foldr generation acc (Sexp.butLast sexp)
+        |> Sexp.addMetaToCar atom
+      where
+        acc =
+          case Sexp.last sexp of
+            -- toss away last %<-... we should likely throw a warning for this
+            Sexp.List [Sexp.Atom (Sexp.A "%<-" _), _name, body] -> body
+            xs -> xs
+        generation body acc =
+          case body of
+            Sexp.List [Sexp.Atom (Sexp.A "%<-" _), name, body] ->
+              Sexp.list
+                [ Sexp.atom "Prelude.>>=",
+                  body,
+                  Sexp.list [Sexp.atom "lambda", Sexp.list [name], acc]
+                ]
+            notBinding ->
+              Sexp.list [Sexp.atom "Prelude.>>", notBinding, acc]
+
+-- | @removePunnedRecords@ - removes the record puns from the syntax to
+-- have an uniform a-list
+-- - BNF input:
+--   + (:record (punned-1) (name-2 body-2) … (punned-n))
+-- - BNF output:
+--   + (:record punned-1 punned-1 name-2 body-2 … punned-n punned-n)
+removePunnedRecords :: Sexp.T -> Sexp.T
+removePunnedRecords xs = Sexp.foldPred xs (== ":record") removePunned
+  where
+    removePunned atom sexp =
+      Sexp.listStar
+        [ Sexp.atom ":record-no-pun",
+          Sexp.foldr f Sexp.Nil sexp
+        ]
+        |> Sexp.addMetaToCar atom
+      where
+        f (Sexp.List [field, bind]) acc =
+          field Sexp.:> bind Sexp.:> acc
+        f (Sexp.List [pun]) acc =
+          pun Sexp.:> pun Sexp.:> acc
+        f _ _ = error "malformed record"
+
+------------------------------------------------------------
+-- Module Pass
+------------------------------------------------------------
+
+-- Update this two fold.
+-- 1. remove the inner combine and make it global
+-- 2. Find a way to handle the cond case
+
+-- | @moduleTransform@ - TODO
+moduleTransform :: Sexp.T -> Sexp.T
+moduleTransform xs = Sexp.foldPred xs (== ":defmodule") moduleToRecord
+  where
+    moduleToRecord atom (name Sexp.:> args Sexp.:> body) =
+      Sexp.list
+        [ Sexp.atom ":defun",
+          name,
+          args,
+          ignoreCond body (\b -> Sexp.foldr combine (generatedRecord b) b)
+        ]
+        |> Sexp.addMetaToCar atom
+    moduleToRecord _ _ = error "malformed record"
+
+moduleLetTransform :: Sexp.T -> Sexp.T
+moduleLetTransform xs = Sexp.foldPred xs (== ":let-mod") moduleToRecord
+  where
+    moduleToRecord atom (name Sexp.:> args Sexp.:> body Sexp.:> rest) =
+      Sexp.list
+        [ Sexp.atom "let",
+          name,
+          args,
+          ignoreCond body (\b -> Sexp.foldr combine (generatedRecord b) b),
+          rest
+        ]
+        |> Sexp.addMetaToCar atom
+    moduleToRecord _ _ = error "malformed record"
+
+----------------------------------------
+-- Module Helpers
+----------------------------------------
+
+combine :: Sexp.T -> Sexp.T -> Sexp.T
```

Comment:
Needs comment at the top, what is this function doing?

Reply:
done